lib/inspec/rule.rb

```diff
@@ -373,6 +373,10 @@ def __apply_waivers
       __waiver_data["skipped_due_to_waiver"] = true
     end
 
+    def is_waiver?(input)
+      input && input.value.is_a?(Hash)
```

Comment:
I've not seen it in action but I think it's possible for an input to be a hash (see https://github.com/inspec/inspec/blob/master/lib/inspec/input.rb#L155 )

This is a little better as it lets users continue to use the control name as an input name as long as they aren't trying to use a hash input.

I think this can only be fixed fully by having InSpec load waivers separately from inputs.

Reply:
I think that's a really good point and we would be warranted to file a separate larger issue for waivers/input loading. 

NOTE: `value` is an attribute (well, actually a method but you know what I mean) on input. So each input has a `value` piece of information you can access. Which is sometimes a string, hash, etc. However, when the input is a waiver, it's *always* a Hash for value. (As __apply_waivers interacts with it like a hash)

The name clash has an unwanted side affect of passing through an input and breaking on this line: https://github.com/inspec/inspec/blob/0d44b0e2818a5ff595bcc2bd9e166139bd0e4d6f/lib/inspec/rule.rb#L343 (because it assumes a Hash for `value` and non-waivers have strings on Value sometimes)

This four-liner blocks non-waiver inputs from trying to be applied. 

I do agree that longer term I would like some more separation, even if there was a type on inputs so I could just call something like `is_waiver?` *directly* on the input :)

I believe this does work for the customer though

Reply:
To make it more straightforward I could pull this method out and we could change the line

`return unless input` (original)

to:

`return unless input && input.value.is_a?(Hash)` 

Which is really what I'm trying to accomplish here. (I just like to name compound conditionals)

Reply:
I think you clearly have investigated this code path more than I have.  

I too typically like to name compound conditionals, but in this case it could introduce some confusion, so let's avoid it.

Naively, I think my order of preference would be:

1) Change waivers to be a specific class we can look for and then look for it here
2) Change the waivers hash to have some sentinel `_is_waiver` key that we can look for if we have a hash here.
3) Use `return unless input && input.value.is_a?(Hash)` with a comments about the situation.

I'm completely cool with (3) for now if (1) and (2) are a bit more than we want to chew off to get this fixed or if they are not good ideas for some other reason.

Reply:
Thank you @stevendanna and @james-stocks for looking over this and your (very helpful) feedback. 

Now that I've had a think on it with this feedback, I think I might revert to no.3, and if it's okay file an issue to move forward on 1&2?

Perhaps working on this issue has helped us uncover the need to class-ify or otherwise identify waivers in the wild which I definitely agree would have made this easier to debug! :) 

If it passes and everyone's happy then I could write up that issue tomorrow, let folks in 4936 know, and squash. (also doing a few small spec updates)

Reply:
Sounds good @Schwad - I think we want to do 1 or 2 at some point because currently inputs and waivers aren't very extensible